core/src/main/scala/org/apache/spark/Partitioner.scala

```diff
@@ -166,7 +169,16 @@ class RangePartitioner[K : Ordering : ClassTag, V](
       // Assume the input partitions are roughly balanced and over-sample a little bit.
       val sampleSizePerPartition = math.ceil(3.0 * sampleSize / rdd.partitions.length).toInt
       val (numItems, sketched) = RangePartitioner.sketch(rdd.map(_._1), sampleSizePerPartition)
-      if (numItems == 0L) {
+      val numSampled = sketched.map(_._3.length).sum
+      if (numItems == 0) {
+        Array.empty
+      }
+      // already got the whole data
+      else if (sampleCacheEnabled && numItems == numSampled) {
+        // get the sampled data
+        sampledArray = sketched.foldLeft(Array.empty[K])((total, sample) => {
+          total ++ sample._3
+        })
         Array.empty
```

Comment:
Returning `Array.empty` here will result in single partition.

Reply:
it may cause errors, I have fixed it